{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "36cc5177",
   "metadata": {},
   "outputs": [],
   "source": [
    "import io # To be able to read strings as files\n",
    "import string # To extract a list of lowercase letters\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Import Biopython for file I/O.\n",
    "import Bio.SeqIO\n",
    "from Bio import AlignIO\n",
    "import Bio.Phylo\n",
    "\n",
    "# Import plotting tools\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "# Read alignment with all species\n",
    "\n",
    "def compute_similarity(seq_1, seq_2):\n",
    "    # Make sure they are the same length. \n",
    "    if len(seq_1) != len(seq_2):\n",
    "        raise ValueError('Sequences must be the same length!')\n",
    "        \n",
    "    # Make both sequences lowercase.\n",
    "    seq_1 = seq_1.lower()\n",
    "    seq_2 = seq_2.lower()\n",
    "        \n",
    "    # Set up counters of length and similarity.\n",
    "    comp_length = 0\n",
    "    num_sim = 0\n",
    "    \n",
    "    # Iterate through each position in the sequences.\n",
    "    for base in range(len(seq_1)):\n",
    "        \n",
    "        # Ensure we are not comparing gaps.\n",
    "        if (seq_1[base] != '-') and (seq_2[base] != '-'):\n",
    "            \n",
    "            # Increase the counter for compared length.\n",
    "            comp_length += 1\n",
    "            \n",
    "            # Compare the two positions.\n",
    "            if seq_1[base] == seq_2[base]:\n",
    "                \n",
    "                # Increase the similarity counter.\n",
    "                num_sim += 1\n",
    "                \n",
    "    # Compute and return the percent similarity.\n",
    "    score = num_sim  / comp_length\n",
    "    return score\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37eca381",
   "metadata": {},
   "source": [
    "### The Neighbor-Joining algorithm\n",
    "##### The procedure begins by assuming that the tree's structure has no hierarchy (the starlike tree). The technique iteratively locates the two nearest nodes and combines them to create a new parental node using the pairwise distance between the nodes as input (like node u in the second schematic). This operation is repeated, adjusting the distance measure as fewer nodes are mismatched. Once two nodes are connected to one another, the algorithm no longer considers them. These nodes can only be linked to the remainder of the tree through the newly generated parental node."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f2778934",
   "metadata": {},
   "outputs": [],
   "source": [
    "def s_matrix(mat):\n",
    "    # Get the number of entries in matrix\n",
    "    n = len(mat)\n",
    "    \n",
    "    # Compute the starlike length\n",
    "    s_star = 1 / (n - 1) * np.sum(np.triu(mat))\n",
    "    # Initialize matrix to save entries\n",
    "    s_mat = np.ones_like(mat) * s_star\n",
    "\n",
    "    # Loop through rows\n",
    "    for i in range(n):\n",
    "        # Loop through columns\n",
    "        for j in range(n):\n",
    "            # Define list of index different from i and j\n",
    "            idx_out = np.arange(n)\n",
    "            idx_out = np.delete(idx_out, [i, j])\n",
    "            \n",
    "            first_term = 1 / (2 * (n - 2)) * np.sum(mat[i, idx_out] + \\\n",
    "                                                    mat[j, idx_out])\n",
    "            \n",
    "            # Second term\n",
    "            second_term = 1 / 2 * mat[i, j]\n",
    "            \n",
    "            reduced_mat = np.delete(mat, [i, j], axis=0)\n",
    "            reduced_mat = np.delete(reduced_mat, [i, j], axis=1)\n",
    "            third_term = 1 / (n - 2) * np.sum(np.triu(reduced_mat))\n",
    "            \n",
    "            # Add result to the s matrix\n",
    "            s_mat[i, j] = first_term + second_term + third_term\n",
    "            \n",
    "    np.fill_diagonal(s_mat, 0)\n",
    "    \n",
    "    return s_mat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "738ca83c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fdc72deb",
   "metadata": {},
   "source": [
    "# For New File output.maf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c9fdd1bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment with 6 rows and 4110 columns\n",
      "GTGTGTTAATCTTACAACCAGAACTCAATTACCCCCTGCATACA...TAT sars2\n",
      "GTGTGTTAATCTTACAACCAGAACTCAATTACCCCCTGCATACA...GCT alpha\n",
      "TTTTACAACCAGAACTCAATTACCCCCTGCATACACTAATTCTT...TAC beta\n",
      "TTTACAAACAGAACTCAATTACCCTCTGCATACACTAATTCTTT...TTT gamma\n",
      "CAGAACTCAATCATACACTAATTCTTTCACACGTGGTGTTTATT...CAG omicron\n",
      "CCCTGCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTG...TTG delta\n"
     ]
    }
   ],
   "source": [
    "aln = AlignIO.read(\"output_all.maf\", \"maf\")\n",
    "print(aln)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a73afdc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['sars2', 'alpha', 'beta', 'gamma', 'omicron', 'delta']\n"
     ]
    }
   ],
   "source": [
    "taxa = []\n",
    "for record in aln:\n",
    "    taxa.append(record.id)\n",
    "\n",
    "print(taxa)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f80de76c",
   "metadata": {},
   "source": [
    "### Implementation \n",
    "##### Although the method appears to involve a lot of unpleasant sums, it is actually rather simple to put into practise. The distance matrix that the method will employ has to be constructed first. In order to do this, I suggest using a function we've already created that compares two sequences pairwise and produces a square symmetric matrix of all the species that are currently accessible.\n",
    "\n",
    "##### We will use 1 - our similarity score as a distance metric since we need something to quantify how far apart two sequences are. This makes sense since, according to the suggested measure, the distance between one species and itself at the diagonal will be zero."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "da8ac7bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distance matrix for all 6 sequences:\n",
      "[[0.         0.68613139 0.71776156 0.74355231 0.71995134 0.74379562]\n",
      " [0.68613139 0.         0.66545012 0.72871046 0.72408759 0.73819951]\n",
      " [0.71776156 0.66545012 0.         0.74136253 0.72092457 0.73746959]\n",
      " [0.74355231 0.72871046 0.74136253 0.         0.74452555 0.73552311]\n",
      " [0.71995134 0.72408759 0.72092457 0.74452555 0.         0.74939173]\n",
      " [0.74379562 0.73819951 0.73746959 0.73552311 0.74939173 0.        ]]\n"
     ]
    }
   ],
   "source": [
    "sim_mat = np.zeros([len(aln), len(aln)])\n",
    "\n",
    "for i, sp1 in enumerate(aln):\n",
    "    # Extract species 1 sequences\n",
    "    seq_1 = sp1.seq\n",
    "    # loop through species in aln\n",
    "    for j, sp2 in enumerate(aln):\n",
    "        # Extract species 2 sequences\n",
    "        seq_2 = sp2.seq\n",
    "        sim_mat[i, j] = 1 - compute_similarity(seq_1, seq_2) ## distance matrix\n",
    "        \n",
    "print('Distance matrix for all {:d} sequences:'.format(len(aln)))\n",
    "print(sim_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d8e85b38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total branch length difference with respect to starlike topology\n",
      "[[0.         2.17305353 2.18381995 2.18287713 2.17542579 2.18166058]\n",
      " [2.17305353 0.         2.16624088 2.18403285 2.18607056 2.18743917]\n",
      " [2.18381995 2.16624088 0.         2.18531022 2.17944039 2.18202555]\n",
      " [2.18287713 2.18403285 2.18531022 0.         2.17740268 2.16721411]\n",
      " [2.17542579 2.18607056 2.17944039 2.17740268 0.         2.17849757]\n",
      " [2.18166058 2.18743917 2.18202555 2.16721411 2.17849757 0.        ]]\n"
     ]
    }
   ],
   "source": [
    "s_mat = s_matrix(sim_mat)\n",
    "print('Total branch length difference with respect to starlike topology')\n",
    "print(s_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "00fbcbc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat = str(s_mat).replace(' [', '').replace('[', '').replace(']', '')\n",
    "with open('matrix.txt', 'w') as f:\n",
    "    f.write(mat)# Save the simulation matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba065d24",
   "metadata": {},
   "source": [
    "### UPGMA Implementation\n",
    "##### The simplest distance-matrix approach is the UPGMA, which uses sequential clustering to create a rooted phylogenetic tree. To construct the distance matrix, all sequences were initially compared using pairwise alignment as we do above.\n",
    "##### It begins by grouping the two taxa that have the least pairwise distance in the distance matrix, given the distance matrix. At the halfway point or distance between them, a node is positioned. The new cluster is then treated as a single taxon to produce a reduced matrix. To build a smaller matrix, the distances between this new composite taxon and all other taxa are determined. Another freshly reduced matrix is produced after repeating the grouping procedure. Once every taxon has been added to the tree, the iteration is complete. The final taxon introduced is regarded as the outgroup that can produce a rooted tree. The fundamental premise of the UPGMA approach is that a molecular clock is operating and that all taxa are equally far from the root and develop at a constant rate. Thus, incorrect tree topologies are frequently generated using UPGMA. However, because of its quick computation speed, it has been widely used in the clustering analysis of DNA microarray data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8edcfec3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "((delta:0.3677615571776156,gamma:0.3677615571776156):0.0026763990267639204,(((beta:0.3327250608272506,alpha:0.3327250608272506):0.018248175182481785,sars2:0.3509732360097324):0.009854014598540128,omicron:0.36082725060827253):0.009610705596106994):0.0\n",
      "\n",
      "      /-delta\n",
      "   /-|\n",
      "  |   \\-gamma\n",
      "  |\n",
      "--|         /-beta\n",
      "  |      /-|\n",
      "  |   /-|   \\-alpha\n",
      "  |  |  |\n",
      "   \\-|   \\-sars2\n",
      "     |\n",
      "      \\-omicron\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "UPGMA - Unweighted pair group method using arithmetic averages\n",
    "Takes as input the distance matrix of species as a numpy array\n",
    "Returns tree either as a dendrogram or in Newick format\n",
    "\"\"\"\n",
    "from ete3 import Tree\n",
    "#%% import modules\n",
    "import numpy as np\n",
    "import itertools\n",
    "\n",
    "#%% class for nodes\n",
    "class Node:\n",
    "    \"\"\"\n",
    "    Data structure to store node of a UPGMA tree\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, left=None, right=None, up_height=0.0, down_height=0.0):\n",
    "        \"\"\"\n",
    "        Creating a node.\n",
    "        For a single taxon, set taxon name as self.left, leave right as none.\n",
    "        For an operational taxonomic unit(OTU) set left and right to child nodes.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        left : default = none, taxon label\n",
    "        right : default = none, taxon label\n",
    "        up_height : float, default = 0.0, dist to parent node, if any\n",
    "        down_height : float, default = 0.0, dist to child node, if any\n",
    "        \"\"\"\n",
    "        self.left = left\n",
    "        self.right = right\n",
    "        self.uh = up_height\n",
    "        self.dh = down_height\n",
    "\n",
    "    def leaves(self) -> list:\n",
    "        \"\"\"\n",
    "        Method to find the taxa under any given node, effectively equivalent to\n",
    "        finding leaves of a binary tree. Only lists original taxa and not OTUs.\n",
    "\n",
    "        Returns a list of node names, not nodes themselves.\n",
    "        \"\"\"\n",
    "        if self == None:\n",
    "            return []\n",
    "        if self.right == None:\n",
    "            return [self.left]\n",
    "        leaves = self.left.leaves() + self.right.leaves()\n",
    "        return leaves\n",
    "\n",
    "    def __len__(self) -> int:\n",
    "        \"\"\"\n",
    "        Method to define len() of a node.\n",
    "\n",
    "        Returns the number of original taxa under any given node.\n",
    "        \"\"\"\n",
    "        return sum(1 for taxa in self.leaves())\n",
    "\n",
    "    def __repr__(self) -> str:\n",
    "        \"\"\"\n",
    "        Method to give readable print output\n",
    "        \"\"\"\n",
    "        return \"-\".join(self.leaves())\n",
    "\n",
    "\n",
    "#%% class for UPGMA\n",
    "class UPGMA:\n",
    "    def __init__(self, dist_matrix: np.ndarray, taxa: list):\n",
    "        \"\"\"\n",
    "        Initialize an UPGMA class.\n",
    "        Takes a nxn distance matrix as input. A list of n taxon id is required\n",
    "        in the same order as the distance matrix row/column\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        dist_matrix : numpy array, distance matrix of species\n",
    "        taxa : list of int or str to identify taxa\n",
    "        \"\"\"\n",
    "        self.distances = dist_matrix\n",
    "        self.taxa = taxa\n",
    "        self.build_tree(self.distances, self.taxa)\n",
    "\n",
    "    def build_tree(self, dist_matrix: np.ndarray, taxa: list) -> Node:\n",
    "        \"\"\"\n",
    "        Method to construct a tree from a given distance matrix and taxa list.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        dist_matrix : np.ndarray of pairwise distances\n",
    "        taxa : list of taxa id. Elements of lists have to be unique\n",
    "\n",
    "        Returns the root node for constructed tree.\n",
    "        \"\"\"\n",
    "        # creating node for each taxa\n",
    "        nodes = list(Node(taxon) for taxon in taxa)\n",
    "\n",
    "        # dictionary row/column id -> node\n",
    "        rc_to_node = dict([i, j] for i, j in enumerate(nodes))\n",
    "        \n",
    "        # dictionary taxa -> row/column id\n",
    "        taxa_to_rc = dict([i, j] for j, i in enumerate(taxa))\n",
    "        \n",
    "        # make copy of dist matrix to work on\n",
    "        work_matrix = dist_matrix\n",
    "        # set all diagonal elements to infinity for ease of finding least distance\n",
    "        work_matrix = np.array(work_matrix, dtype=float)\n",
    "        np.fill_diagonal(work_matrix, np.inf)\n",
    "\n",
    "        # loop\n",
    "        while len(nodes) > 1:\n",
    "            # finding (row, col) of least dist\n",
    "            least_id = np.unravel_index(work_matrix.argmin(), work_matrix.shape, \"C\")\n",
    "            least_dist = work_matrix[least_id[0], least_id[1]]\n",
    "            # nodes corresponding to (row, col)\n",
    "            node1, node2 = rc_to_node[least_id[0]], rc_to_node[least_id[1]]\n",
    "            \n",
    "            # add OTU with node1 and node2 as children. set heights of nodes\n",
    "            new_node = Node(node2, node1)\n",
    "            nodes.append(new_node)\n",
    "            node1.uh = least_dist / 2 - node1.dh\n",
    "            node2.uh = least_dist / 2 - node2.dh\n",
    "            new_node.dh = least_dist / 2\n",
    "            nodes.remove(node1)\n",
    "            nodes.remove(node2)\n",
    "           \n",
    "            # create new working distance matrix\n",
    "            work_matrix = self.update_distance(dist_matrix, nodes, taxa_to_rc)\n",
    "            \n",
    "            # update row/col id -> node dictionary\n",
    "            rc_to_node = dict([i, j] for i, j in enumerate(nodes))\n",
    "        # set tree to root\n",
    "        self.tree = nodes[0]\n",
    "\n",
    "    def update_distance(\n",
    "        self, dist_matrix: np.ndarray, nodes: list, taxa_to_rc: dict\n",
    "    ) -> np.ndarray:\n",
    "        \"\"\"\n",
    "        Method to make a new distance matrix with newer node list.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        dist_matrix : np.ndarray of pairwise distances for all taxa\n",
    "        nodes : list of updated nodes\n",
    "        taxa_to_rc : dict for taxa -> row/col id\n",
    "\n",
    "        Returns np.ndarray of pairwise distances for updated nodes\n",
    "        \"\"\"\n",
    "        # dictionary for node -> row/col id\n",
    "        node_to_rc = dict([i, j] for j, i in enumerate(nodes))\n",
    "        \n",
    "        rc = len(nodes)\n",
    "        new_dist_matrix = np.zeros((rc, rc), dtype=float)\n",
    "        for node1 in nodes:\n",
    "            row = node_to_rc[node1]\n",
    "            for node2 in nodes:\n",
    "                node_pairs = list(itertools.product(node1.leaves(), node2.leaves()))\n",
    "                col = node_to_rc[node2]\n",
    "                new_dist_matrix[row, col] = sum(\n",
    "                    dist_matrix[taxa_to_rc[i], taxa_to_rc[j]] for i, j in node_pairs\n",
    "                ) / len(node_pairs)\n",
    "        np.fill_diagonal(new_dist_matrix, np.inf)\n",
    "        return new_dist_matrix\n",
    "\n",
    "\n",
    "def tree_to_newick(t) -> str:\n",
    "    \"\"\"\n",
    "    Function to convert tree to Newick, slightly modified form of the tree.py version.\n",
    "    Takes the root node of an UPGMA tree as input\n",
    "    \"\"\"\n",
    "    if t.right == None:\n",
    "        return t.left + \":\" + str(t.uh)\n",
    "    else:\n",
    "        return (\n",
    "            \"(\"\n",
    "            + \",\".join([tree_to_newick(x) for x in [t.left, t.right]])\n",
    "            + \"):\"\n",
    "            + str(t.uh)\n",
    "        )\n",
    "\n",
    "def tree_to_newick(t) -> str:\n",
    "    \"\"\"\n",
    "    Function to convert tree to Newick, slightly modified form of the tree.py version.\n",
    "    Takes the root node of an UPGMA tree as input\n",
    "    \"\"\"\n",
    "    if t.right == None:\n",
    "        return t.left + \":\" + str(t.uh)\n",
    "    else:\n",
    "        return (\n",
    "            \"(\"\n",
    "            + \",\".join([tree_to_newick(x) for x in [t.left, t.right]])\n",
    "            + \"):\"\n",
    "            + str(t.uh)\n",
    "        )\n",
    "#%% main\n",
    "if __name__ == \"__main__\":\n",
    "    # data from table 3 of Fitch and Margoliash, Construction of Phylogenetic trees\n",
    "    #taxa = list(map(chr, range(97,97+len(aln))))\n",
    "    x = UPGMA(sim_mat, taxa).tree\n",
    "    print(tree_to_newick(x))\n",
    "    t=Tree(tree_to_newick(x)+\";\")\n",
    "    print(t)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
